docs/proposals/01-firmware-update.md

```diff
@@ -0,0 +1,381 @@
+
+- [API server](#api-server)
+- [scheduler](#scheduler)
+- [job runner](#job-runner)
+- [update/scan/discovery jobs](#updatescandiscovery-job) (concrete implementations)
+
+Update service MUST be extensible in part of the possibility of using plugins for update strategy.
+At the same time, the only update strategy can be enabled for a particular update service instance.
+Update service MAY also include the component to download and store the firmwares.
+
+#### API server
+
+API server exposes update service endpoints and forwards incoming requests to the scheduler.
+It MUST expose the following endpoints:
+
+- Update(UpdateRequest) UpdateResponse;
+  - `UpdateRequest` MUST contain the reference to concrete `Server` object and the list of the firmware-version to be installed.
+  - `UpdateResponse` MUST contain the status of the request with error code if any.
+- Scan(ScanRequest) ScanResponse;
+  - `ScanRequest` MUST contain the reference to concrete `Server` object.
+  - `ScanResponse` MUST contain the status of the request with error code if any.
+- Discover(DiscoverRequest) DiscoverResponse;
+  - `DiscoverRequest` MUST contain the reference to concrete `DiscoveredFirmware` object.
+  - `DiscoverResponse` MUST contain the status of the request with error code if any.
+- UpdateServer(UpdateServerRequest) UpdateServerResponse;
+  - `UpdateServerRequest` MUST contain the reference to concrete `Server` object and the list of installed firmware-versions.
+  This endpoint MUST be used by update or scan jobs after a task if finished to send results and invoke the object's update.
+  - `UpdateServerResponse` MUST contain the status of the request with error code if any.
+- UpdateDiscoveredFirmware(UpdateDiscoveredFirmwareRequest) UpdateDiscoveredFirmwareResponse;
+  - `UpdateDiscoveredFirmwareRequest` MUST contain the reference to concrete `DiscoveredFirmware` object and the list of discovered firmware-versions.
+  This endpoint MUST be used by discovery jobs after a task is finished to send results and invoke the object's update.
+  - `UpdateDiscoveredFirmwareResponse` MUST contain the status of the request with error code if any.
+
+Depending on the type of the request, it SHOULD be forwarded to the corresponding scheduler's queue.
+
+#### Scheduler
+
+Scheduler is a component of the update service that is responsible for scheduling jobs:
+
+- it MUST NOT allow running several jobs on the same target server simultaneously;
+- it MAY discard incoming update or scan requests if the same jobs targeting the same server are already scheduled;
+- it MAY discard incoming discovery requests if the job targeting the same manufacturer-model pair is already scheduled;
+- it MUST have a mechanism to limit the number of parallel jobs;
+- it MUST have a mechanism to limit the job queue length;
+
+Scheduler SHOULD have an embedded job runner component corresponding to the update strategy defined on the update service application start.
+
+#### Job runner
```

Comment:
Yep, the idea is to leverage kubernetes jobs. But the application which job would execute would depend on the update strategy, for instance:
- in case of using vendor's CLI tool, the app inside job will execute that tool
- in case of using prepared boot image the app inside job will create boot config and patch server object to use it
- etc.

Reply:
Understood.